src/app/icd/ICDHandler.h

```diff
@@ -0,0 +1,90 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines objects for a CHIP CheckInMessage unsolicited
+ *      handler
+ *
+ */
+
+#pragma once
+
+#include <lib/core/CHIPCore.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/DLLUtil.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <protocols/icd/Constants.h>
+#include <protocols/secure_channel/CheckinMessage.h>
+
+namespace chip {
+namespace app {
+namespace CheckIn {
+
+class CheckInMessageHandler : public Messaging::ExchangeDelegate, public Messaging::UnsolicitedMessageHandler
+{
+    class Callback
+    {
+    public:
+        virtual ~Callback() = default;
+
+        // TODO : Include the callback message from ICDClientManagement
+
+        /*
+         * Upon processing of a CheckInMessage, this method is invoked to dispatch the result
+         * to the right client-side handler provided by the application.
+         */
+        // virtual void DispatchCheckIn(CommandHandler & apCommandObj, const ConcreteCommandPath & aCommandPath,
+        //                              TLV::TLVReader & apPayload) = 0;
+    };
+
+public:
+    /**
+     * @brief Retrieve the singleton CheckIn handler
+     *
+     *  @return  A pointer to the shared CheckIn handler
+     *
+     */
+    static CheckInMessageHandler * GetInstance(void);
+
+    CHIP_ERROR Init(Messaging::ExchangeManager * exchangeManager);
+    void Shutdown();
+
+    Messaging::ExchangeManager * GetExchangeManager(void) const { return mExchangeManager; }
+
+protected:
+    // ExchangeDelegate
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                 System::PacketBufferHandle && payload) override;
+    void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
```

Comment:
We should really not make this pure-virtual on ExchangeDelegate... It makes no sense to have it here.  Followup for that, please?

Reply:
I'll follow-up on that.